doc/code/qml_hf.rst

```diff
@@ -1,7 +1,144 @@
+performs self-consistent-field iterations to obtain a set of optimized molecular orbital
+coefficients. These coefficients and the computed integrals over basis functions are used to
+construct the one- and two-body electron integrals in the molecular orbital basis which can be
+used to generate a differentiable second-quantized Hamiltonian in the fermionic and qubit basis.
+
+The following code shows the construction of the Hamiltonian for the hydrogen molecule where the
+geometry of the molecule and the basis set parameters are all differentiable.
+
+.. code-block:: python3
+
+    import pennylane as qml
+    from pennylane import numpy as np
+
+    symbols = ["H", "H"]
+    geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 2.0]], requires_grad=True)
+
+    # alpha and coeff are the exponents and contraction coefficients of the Gaussian functions
+    alpha = np.array([[3.42525091, 0.62391373, 0.1688554],
+                      [3.42525091, 0.62391373, 0.1688554]], requires_grad = True)
+    coeff = np.array([[0.15432897, 0.53532814, 0.44463454],
+                      [0.15432897, 0.53532814, 0.44463454]], requires_grad = True)
+
+We create a molecule object with differentiable atomic coordinates and basis set parameters and then
+construct the Hamiltonian.
+
+.. code-block:: python3
+
+    mol = qml.hf.Molecule(symbols, geometry, alpha=alpha, coeff=coeff)
+    args_mol = [geometry, alpha, coeff] # initial values of the differentiable parameters
+
+    hamiltonian = qml.hf.generate_hamiltonian(mol)(*args_mol)
+
+    >>> print(hamiltonian)
+      ((-0.3596823592263728+0j)) [I0]
+    + ((-0.11496335836149135+0j)) [Z3]
+    + ((-0.1149633583614913+0j)) [Z2]
+    + ((0.1308241430373499+0j)) [Z0]
+    + ((0.1308241430373499+0j)) [Z1]
+    + ((0.10316898251626505+0j)) [Z0 Z2]
+    + ((0.10316898251626505+0j)) [Z1 Z3]
+    + ((0.1532995999427217+0j)) [Z0 Z3]
+    + ((0.1532995999427217+0j)) [Z2 Z1]
+    + ((0.1540549585580985+0j)) [Z0 Z1]
+    + ((0.1609686663985837+0j)) [Z3 Z2]
+    + ((-0.05013061742645664+0j)) [Y0 X2 X3 Y1]
+    + ((-0.05013061742645664+0j)) [X0 Y2 Y3 X1]
+    + ((0.05013061742645664+0j)) [Y0 X2 Y3 X1]
+    + ((0.05013061742645664+0j)) [X0 Y2 X3 Y1]
+
+The generated Hamiltonian can be used in a circuit where the molecular geometry, the basis set
+parameters and the circuit parameters are optimized simultaneously.
+
+.. code-block:: python3
+
+    import autograd
+
+    dev = qml.device("default.qubit", wires=4)
+    hf_state = np.array([1, 1, 0, 0])
+    params = [np.array([0.0], requires_grad=True)] # initial values of the circuit parameters
+
+    def generate_circuit(mol):
+        @qml.qnode(dev)
+        def circuit(*args):
+            qml.BasisState(hf_state, wires=[0, 1, 2, 3])
+            qml.DoubleExcitation(*args[0][0], wires=[0, 1, 2, 3])
+            return qml.expval(qml.hf.generate_hamiltonian(mol)(*args[1:]))
+        return circuit
+
+    for n in range(10): # geometry and parameter optimization loop
```

Comment:
Could we add a short linker sentence here? I find breaking up large codeblocks helps with comprehension sometimes.

```suggestion
        return circuit

Now that the circuit is defined, we can create a geometry
and parameter optimization loop:

.. code-block:: python3

    for n in range(10):
```